swupd/manifest.go

```diff
@@ -467,6 +468,74 @@ func (m *Manifest) newDeleted(df *File) {
 	m.Files = append(m.Files, df)
 }
 
+// TODO: Consolidate this function with linkPeersAndChange.
+
+// linkDeltaPeers sets the DeltaPeer of the files in newManifest that have the corresponding files
+// in oldManifest.
+func linkDeltaPeers(c *config, oldManifest, newManifest *Manifest) error {
+	var added, removed []*File
+	newIndex := 0
+	oldIndex := 0
+
+	for newIndex < len(newManifest.Files) && oldIndex < len(oldManifest.Files) {
+		nf := newManifest.Files[newIndex]
+		of := oldManifest.Files[oldIndex]
+
+		switch {
+		case nf.Name < of.Name:
+			// Old file that is not in new manifest, no chance of delta.
+			removed = append(removed, of)
+			newIndex++
+
+		case nf.Name > of.Name:
+			// New file that wasn't present before, no chance of delta.
+			added = append(added, nf)
+			oldIndex++
+
+		default:
+			newIndex++
+			oldIndex++
+
+			// Matching names, we can have a delta if pass all the
+			// requirements below.
+			if nf.Version <= of.Version {
+				continue
+			}
+
+			if nf.Hash == of.Hash {
+				continue
+			}
+
+			if !nf.Present() || !nf.Present() {
+				continue
+			}
+
+			if nf.Type != of.Type || nf.Type != TypeFile {
```

Comment:
Optional: `nf.Type != TypeFile` is much more likely to be `true` than `nf.Type != of.Type` since type changes are rare. If you move the second conditional before the first you will be doing much less comparisons for the full update.

Reply:
Done.